Generic abstractions for virtio devices and the MMIO transport

alexandruag opened the pull request:
This PR proposes generic abstractions for virtio devices and the MMIO transport in particular (the aim is to add PCI support as well at some point). We first redefine the `VirtioDevice` trait, which contains methods required to support interactions that happen at the transport level. The following commit defines `VirtioMmioDevice`, which provides default implementations (based on the functionality from `VirtioDevice`) for reads and writes to the registers/configuration space of a virtio device that uses the MMIO transport. Finally, the last commit introduces the `VirtioConfig`, `WithVirtioConfig`, and `WithDeviceOps` helper abstractions, which are used to provide automatic implementations of the previous traits.

There are more details present in the code comments. This [commit](https://github.com/alexandruag/vm-virtio/commit/32f26359c6fa5996f62ae767115c793f43ab78af#diff-060bce9da11057f730f84bbc0e029eb859f54cc5e05b05bed48f5a8f1d4df524) contains a simple example of how the previous pieces fit together to simplify the implementation of virtio-over-MMIO functionality for a potential device. I'm working on a more complete example of an actual working device and will add it here as well once it's available. Looking forward to any feedback!

alexandruag commented:
[Here](https://github.com/alexandruag/vmm-reference/blob/block_rfc/src/devices/src/virtio/block/device.rs)'s a better example of how the previously mentioned abstractions can be used as part of a block device implementation.

aghecenco commented on src/devices/mmio.rs:
```diff
@@ -0,0 +1,199 @@
+            }
+
+            // TODO: The standard specifies that configuration space size is device specific, so
+            // we might want to express that via the trait instead of hard coding the current
```
+1. It can have a const fn and default implementation, not a big change, and would be neater.

alexandruag replied:
On second though, the effective size of the configuration space is actually variable depending on selected features, so I would keep the static upper bound and update TODO for now to think a bit more about this. Ultimately, each specific implementation of `read/write_config` takes care of validations, etc.

aghecenco commented on src/devices/mmio.rs:
```diff
@@ -0,0 +1,199 @@
+            // we might want to express that via the trait instead of hard coding the current
+            // arbitrary ceiling.
+            // It's ok to use `as` here because `offset` always fits into an `usize` in this case.
+            0x100..=0xfff => self.read_config(offset as usize - 0x100, data),
```
According to the spec: `Device-specific configuration space starts at the offset 0x100 and is accessed with byte alignment.`
Admittedly, `byte alignment` is a weird phrase but looking next door, qemu interprets this as `only byte, word or dword accesses are allowed` ([code](https://github.com/qemu/qemu/blob/c8e5c4b246584da36694a3c259a7dbb8a7e7b1f3/hw/virtio/virtio-mmio.c#L113)). I think this makes sense and we should do the same check, WDYT?

rbradford replied:
The byte alignment here means you're not restricted to only accessing it at a word,dword or qword granularity. i.e. you can access it however you like so there is no check you can make.

slp commented:
I like both the design and the implementation. :+1:

alexandruag commented:
Thanks for the comments everyone! I'll address them and turn this into a regular PR over the following days.

andreeaflorescu commented on src/lib.rs:
```diff
@@ -14,10 +14,10 @@ extern crate log;
 mod queue;
 
-pub use self::device::*;
+pub use self::devices::*;
```
I think that queues and devices are conceptually different and can be used independently (at least the queues). For this reason, what do you say about exporting the modules directly instead of `pub use`?

My general thinking when using flattened vs non-flattened exports is that flattened exports are to be used when the modules are there just to provide a structural separation, and not a logical one.

**Structural separation**: If everything would be in one module, than the module would be to big, and thus the code becomes hard to follow, hence we are splitting it in multiple modules. This artificial separation is hidden from the end user by exporting everything flattened.
**Logical separation**: separate concepts are split between logical modules.

andreeaflorescu commented on src/devices/mmio.rs:
```diff
@@ -0,0 +1,199 @@
+// a `VirtioDevice`, provided the status check is successful.
+// TODO: This function and its uses will likely have to be updated when we start offering
+// packed virtqueue support as well.
+fn update_queue_field<M, D, F>(device: &mut D, f: F)
```
I deeply dislike these kind of functions that don't belong to any object :(

andreeaflorescu reviewed (commented):
I just looked briefly over the interfaces. I'll do an in-depth review of the implementation when this is not longer an RFC.

I find the `WithDriverSelection` a bit too strange. I believe we can work on improving the readability of that code by providing a better abstraction (that is in case I understood correctly what you're trying to achieve with it).

Otherwise, the PR LGTM.

alexandruag commented on src/devices/virtio_config.rs:
```diff
@@ -0,0 +1,248 @@
+// this in favor of adding accessors.
+pub struct VirtioConfig<M: GuestAddressSpace> {
+    /// The set of features exposed by the device.
+    pub device_features: u64,
```
Yeah so like I mentioned in a previous comment, the driver can write a full 32-bit value when selecting a page, but then all the feature bits defined by the standard so far appear to fit within a `u64`. I'll use the simplification mentioned there if it sounds good.

alexandruag commented on src/devices/virtio_config.rs:
```diff
@@ -0,0 +1,248 @@
+
+/// Another helper trait, which (in conjunction with `WithVirtioConfig`) is used to provide
+/// automatic implementations of `VirtioDevice` and `WithDriverSelect`.
+pub trait WithDeviceOps {
```
Yeah so I don't know which name is best here, but at the same time I don't know if `DeviceStateUpdate` fits that well. We can think a bit more and just go with something in the end.

alexandruag commented on src/devices/virtio_config.rs:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    fn num_feature_pages(&self) -> u32 {
+        // We use an `u64` to keep track of features within `VirtioConfig`.
```
This method might disappear, depending on the discussion around the previous comment about removing it.

alexandruag commented:
I've simplified `devices_features`/`driver_features` handling a bit (as per one of the previous comments) and added unit tests in a new, separate commit. I also removed the draft status from the PR. Thanks for all the comments so far and please take another look.

alexandruag commented:
Did another push that renames some of the helper traits in the fourth commit, uses `BorrowMut` to abstract access to a `VirtioConfig` object, and we're also setting the `event_idx` queue flag as part of feature acknowledgement for extra convenience.

aghecenco commented on src/device/mmio.rs:
```diff
@@ -0,0 +1,355 @@
+                        0 => self.device_features() as u32,
+                        1 => (self.device_features() >> 32) as u32,
+                        // No device features defined beyond the first two pages.
+                        _ => 0,
```
This can be tested too fairly easily if the `Dummy` device does `d.set_device_features_select(2)` before an MMIO read from offset `0x10`

aghecenco commented on src/device/mmio.rs:
```diff
@@ -0,0 +1,355 @@
+                    0x30 => self.set_queue_select(v as u16),
+                    0x38 => update_queue_field(self, |q| q.size = v as u16),
+                    0x44 => update_queue_field(self, |q| q.ready = v == 1),
+                    0x50 => self.queue_notify(v),
```
This is also rather easily testable if you override `fn queue_notify` for `Dummy` and force a write here.

aghecenco commented on src/device/mod.rs:
```diff
@@ -0,0 +1,289 @@
+                // Set the appropriate configuration flag for all queues if we offered the
+                //`VIRTIO_F_RING_EVENT_IDX` feature and the driver acknowledged it.
+                if self.driver_features() & (1 << VIRTIO_F_RING_EVENT_IDX) != 0 {
+                    for i in 0..self.num_queues() {
```
I think there's a way to test this too with `Dummy` though I don't have the specifics.

aghecenco reviewed (approved):
I have some final nits and minor test enhancements that don't look too hard to add, but I think the PR is good to go as-is. Lack of proper documentation is my main note here, but that can wait until the crate is ready for publishing.

slp commented:
LGTM. I think it can be merged once is rebased.

alexandruag commented on src/device/virtio_config.rs:
```diff
@@ -0,0 +1,227 @@
+            return;
+        }
+
+        // TODO: Are partial reads ok?
```
Sure, will add an issue immediately after this gets merged in the current form.

alexandruag reviewed (commented):
Addressed the latest comments and did a rebase.

jiangliu commented on src/device/mod.rs:
```diff
@@ -0,0 +1,323 @@
+    /// Indicates that something went wrong in the guest, and it has given up on the device.
+    /// This could be an internal error, or the driver didn’t like the device for some reason,
+    /// or even a fatal error during device operation.
+    pub const FAILED: u8 = 128;
```
Could we sort these consts by value?

alexandruag replied:
They were initially written down in the order they are presented in the standard (i.e. in section 2.1), but we can def reorder them if it improves clarity & readability.

alexandruag commented on .buildkite/custom-pipeline.yml:
```diff
@@ -0,0 +1,52 @@
+      os: linux
+    plugins:
+      - docker#v3.0.1:
+          image: "rustvmm/dev:v6"
```
Yeah it makes sense to do so; we can piggy back the update on another PR.

andreeaflorescu replied:
I've opened an issue to update the rust-vmm-ci as well as it is super old: https://github.com/rust-vmm/vm-virtio/issues/34

I'll add a comment to also update the container version in the custom-pipeline.
